File: problems/150-evaluate-reverse-polish-notation/csharp/150_evaluate_reverse_polish_notation.cs
using System.Collections.Generic;
using System;

public static class Program
{
  public static void Main(string[] args)
  {
    Solution s = new Solution();
    Console.WriteLine(s.EvalRPN(new string[] { "2", "1", "+", "3", "*" }));
    Console.WriteLine(s.EvalRPN(new string[] { "4", "13", "5", "/", "+" }));
    Console.WriteLine(s.EvalRPN(new string[] { "10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+" }));
  }
}

public class Solution
{
  public int EvalRPN(string[] tokens)
  {
    Stack<int> stack = new Stack<int>();
    HashSet<string> operators = new HashSet<string> { "+", "-", "*", "/" };

    foreach (string token in tokens)
    {
      if (!operators.Contains(token))
        stack.Push(int.Parse(token));
      else
      {
        int b = stack.Pop();
        int a = stack.Pop();
        switch (token)
        {
          case "+": stack.Push(a + b); break;
          case "-": stack.Push(a - b); break;
          case "*": stack.Push(a * b); break;
          case "/": stack.Push(a / b); break;
        }
      }
    }

    return stack.Pop();
  }
}

File: problems/150-evaluate-reverse-polish-notation/python/150_evaluate_reverse_polish_notation.py
from typing import List, Set


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack: List[int] = []
        operators: Set[str] = {"+", "-", "*", "/"}

        for token in tokens:
            if token not in operators:
                stack.append(int(token))
            else:
                b: int = stack.pop()
                a: int = stack.pop()
                match token:
                    case "+":
                        stack.append(int(a + b))
                    case "-":
                        stack.append(int(a - b))
                    case "*":
                        stack.append(int(a * b))
                    case "/":
                        stack.append(int(a / b))

        return stack.pop()


def test(tokens: List[str]) -> None:
    result: int = Solution().evalRPN(tokens)
    print(f"Value for these tokens is {result}! \n -> tokens: {tokens}")


if __name__ == "__main__":
    print()
    test(["2", "1", "+", "3", "*"])
    print()
    test(["4", "13", "5", "/", "+"])
    print()
    test(["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"])
    print()
